Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp

```diff
@@ -1632,38 +1808,40 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes(GridLayoutState& gridL
 
 void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizesMasonry(GridLayoutState& gridLayoutState)
 {
-    if (m_strategy->isComputingSizeContainment()) {
+    if (m_strategy->isComputingSizeContainment() || !m_grid.hasGridItems()) {
         handleInfinityGrowthLimit();
         return;
     }
 
-    Vector<GridTrack>& allTracks = tracks(m_direction);
-    Vector<GridItemWithSpan> itemsSortedByIncreasingSpan;
-    Vector<GridItemWithSpan> itemsCrossingFlexibleTracks;
-    SingleThreadWeakHashSet<RenderBox> itemsSet;
-    MasonryIndefiniteItems masonryIndefiniteItems;
+    StdMap<unsigned, MasonryMinMaxTrackSize> indefiniteTrackSizes;
+    StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItems;
+    StdMultimap<unsigned, GridItemWithSpan> definiteMultiTrackItemsWithFlex;
 
-    if (m_grid.hasGridItems()) {
-        computeIndefiniteItemsForMasonry(masonryIndefiniteItems, gridLayoutState);
+    // Compute Span Sizes
+    computeIndefiniteItemsForMasonry(indefiniteTrackSizes, definiteMultiTrackItems, definiteMultiTrackItemsWithFlex, gridLayoutState);
+
+    // Single Track Spans
+    auto& allTracks = tracks(m_direction);
 
+    if (indefiniteTrackSizes.contains(1)) {
+        auto& singleTrackSpanSize = indefiniteTrackSizes.find(1)->second;
```

Comment:
You can probably rewrite this to avoid the redudant lookups by moving the find into the conditional